File: LeetCode/0076.Minimum-Window-Substring.cpp
// 76. Minimum Window Substring

class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char,int> mps;
        unordered_map<char,int> mpt;
        
        for(auto i : t) {
            mpt[i]++;
        }
        
        int i = 0;
        int j = 0;
        int n = s.length();
        int cnt = 0;
        int dcnt = t.length();
        string ans = "";
        
        while(true) {
            bool f1 = false;
            bool f2 = false;
            
            while(i<n && cnt<dcnt) {
                if(mpt.find(s[i]) != mpt.end()) {
                    mps[s[i]]++;
                    if(mps[s[i]] <= mpt[s[i]]) {
                        cnt++;
                    }
                }
                f1 = true;
                i++;
            }
            
            while(j<i && cnt == dcnt) {
                string temp = s.substr(j,i-j);
                if(ans.length() == 0 || temp.length()<ans.length()) {
                    ans = temp;
                }
                if(mpt.find(s[j]) != mpt.end()) {
                    if(mps[s[j]] == 1) {
                        mps.erase(s[j]);
                    }
                    else {
                        mps[s[j]]--;
                    }
                    if(mps[s[j]] < mpt[s[j]]) {
                        cnt--;
                    }
                }
                f2 = true;
                j++;
            }
            
            if(f1 == false && f2 == false) {
                break;
            }
        }
        return ans;
    }
};

File: LeetCode/0076-Minimum-Window-Substring.java
// 76. Minimum Window Substring

class Solution {
    boolean isIncludeT(Map<Character,Integer> mp) {
        for (int f : mp.values()) {
            if (f > 0) return false;
        }
        return true;
    }

    public String minWindow(String s, String t) {
        int n = s.length();

        Map<Character,Integer> mp = new HashMap<>();
        for (char ch : t.toCharArray()) mp.put(ch, mp.getOrDefault(ch, 0) + 1);

        int l = 0, r = 0, mn = Integer.MAX_VALUE;
        String ans = "";
        while (r < n) {
            char ch = s.charAt(r);
            if (mp.containsKey(ch)) mp.put(ch, mp.get(ch) - 1);
            while (isIncludeT(mp)) {
                if (r - l + 1 < mn) {
                    mn = r - l + 1;
                    ans = s.substring(l, l + mn);
                }
                ch = s.charAt(l);
                if (mp.containsKey(ch)) mp.put(ch, mp.get(ch) + 1);
                l++;
            }
            r++;
        }
        return ans;
    }
}
